pkg/controller/job/job_controller.go

```diff
@@ -508,39 +502,50 @@ func (jm *Controller) deleteJob(logger klog.Logger, obj interface{}) {
 	}
 }
 
-// obj could be an *batch.Job, or a DeletionFinalStateUnknown marker item,
-// immediate tells the controller to update the status right away, and should
-// happen ONLY when there was a successful pod run.
-func (jm *Controller) enqueueController(logger klog.Logger, obj interface{}, immediate bool) {
-	jm.enqueueControllerDelayed(logger, obj, immediate, 0)
+// enqueueSyncJobImmediately tells the Job controller to invoke syncJob
+// immediately.
+// It is only used for Job events (creation, deletion, spec update).
+// obj could be an *batch.Job, or a DeletionFinalStateUnknown marker item.
+func (jm *Controller) enqueueSyncJobImmediately(logger klog.Logger, obj interface{}) {
+	jm.enqueueSyncJobInternal(logger, obj, 0)
+}
+
+// enqueueSyncJobBatched tells the controller to invoke syncJob with a
+// constant batching delay.
+// It is used for:
+// - Pod events (creation, deletion, update)
+// - Job status update
+// - Job status error due to a conflict
+// obj could be an *batch.Job, or a DeletionFinalStateUnknown marker item.
+func (jm *Controller) enqueueSyncJobBatched(logger klog.Logger, obj interface{}) {
+	jm.enqueueSyncJobInternal(logger, obj, jm.syncJobBatchPeriod)
 }
 
-func (jm *Controller) enqueueControllerPodUpdate(logger klog.Logger, obj interface{}, immediate bool) {
-	jm.enqueueControllerDelayed(logger, obj, immediate, jm.podUpdateBatchPeriod)
+// enqueueSyncJobWithBackoff tells the controller to invoke syncJob with a
+// custom delay, but not smaller than the batching delay.
+// It is used when pod recreations are delayed due to pod failures.
+// obj could be an *batch.Job, or a DeletionFinalStateUnknown marker item.
+func (jm *Controller) enqueueSyncJobWithBackoff(logger klog.Logger, obj interface{}, delay time.Duration) {
+	if delay < jm.syncJobBatchPeriod {
+		delay = jm.syncJobBatchPeriod
+	}
+	jm.enqueueSyncJobInternal(logger, obj, delay)
 }
 
-func (jm *Controller) enqueueControllerDelayed(logger klog.Logger, obj interface{}, immediate bool, delay time.Duration) {
+func (jm *Controller) enqueueSyncJobInternal(logger klog.Logger, obj interface{}, delay time.Duration) {
 	key, err := controller.KeyFunc(obj)
 	if err != nil {
 		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
 		return
 	}
-
-	backoff := delay
```

Comment:
We should still have this backoff for API failures.

Reply:
I'm not sure we have a good use case to reimplement the mechanism that is already present inside the delaying queues, as we already use [AddRateLimited]( https://github.com/kubernetes/kubernetes/blob/fa78f28f0a011e9922cf12145dd602c5a63981e8/pkg/controller/job/job_controller.go#L576).

The only place where this is used is when handling the conflict [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L897-L902). 
However, I'm confused by that, because the comment says:
```
// we probably have a stale informer cache so don't return an error to avoid backoff
```
However, we queue with backoff. 

So following the comment I enqueued just with delay. However, we could also just return error, which would result in calling `AddRateLimited`, so adding backoff.

I have also looked at endpointslice & daemonset controllers, and they don't do a special case for conflicts. So, maybe it is better to just remove the special case and return error in that case?

Reply:
> I'm not sure we have a good use case to reimplement the mechanism that is already present inside the delaying queues, as we already use AddRateLimited [here](https://github.com/kubernetes/kubernetes/blob/fa78f28f0a011e9922cf12145dd602c5a63981e8/pkg/controller/job/job_controller.go#L576).

Oh ok, that should be enough.

> I have also looked at endpointslice & daemonset controllers, and they don't do a special case for conflicts. So, maybe it is better to just remove the special case and return error in that case?

Uhm... yeah, perhaps better to be consistent.

Reply:
And I think lowering the values will also be beneficial for the features using `ReplacementPolicy: Failed`.

Reply:
FYI @soltysh @wojtek-t in case you have an opinion

Reply:
decreasing start to 1s sounds fine

I would leave the maxbackoff larger than 10s, 1m?

Reply:
Talked to @wojtek-t and he is fine with lowering, but not below `1s`.

Reply:
I like `1m`, so that the change in behavior is smaller.

Reply:
@alculquicondor would you like me to adjust the constants in this PR, or a dedicated one?

Reply:
Here seems fine